src/backend/distributed/transaction/multi_transaction.c

```diff
@@ -0,0 +1,1907 @@
+
+	FinishConnectionEstablishment(connection);
+	return connection;
+}
+
+
+static void
+AdjustRemoteTransactionStates(List *connectionList)
+{
+	ListCell *connectionCell = NULL;
+
+	if (!InCoordinatedTransaction())
+		return;
+	foreach(connectionCell, connectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+
+		Assert(PQtransactionStatus(connection->conn) != PQTRANS_ACTIVE);
+
+		if ((connection->transactionState & MULTI_TRANSACTION_STATE_FAILED) ||
+			(connection->transactionState & MULTI_TRANSACTION_STATE_STARTING) ||
+			(connection->transactionState & MULTI_TRANSACTION_STATE_STARTED))
+		{
+			continue;
+		}
+
+		if (PQtransactionStatus(connection->conn) != PQTRANS_INTRANS)
+		{
+			/*
+			 * Check whether we're right now allowed to start new client
+			 * transaction.  FIXME: This likely can be removed soon.
+			 */
+			if (XactModificationLevel > XACT_MODIFICATION_NONE)
+			{
+				ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+								errmsg("cannot open new connections after the first modification "
+									   "command within a transaction")));
+			}
+
+			/* FIXME: specify isolation level etc. */
+			if (!SendRemoteCommand(connection, "BEGIN;"))
+			{
+				ReportConnectionError(connection, WARNING);
+				connection->transactionState |= MULTI_TRANSACTION_STATE_FAILED;
+			}
+			else
+			{
+				connection->transactionState |= MULTI_TRANSACTION_STATE_STARTING;
+			}
+		}
+	}
+
+
+	foreach(connectionCell, connectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		PGresult *result = NULL;
+
+		if (connection->transactionState & MULTI_TRANSACTION_STATE_FAILED)
+		{
+			continue;
+		}
+
+		if (!(connection->transactionState & MULTI_TRANSACTION_STATE_STARTING))
+		{
+			continue;
+		}
+
+		result = PQgetResult(connection->conn);
+		if (!IsResponseOK(result))
+		{
+			connection->transactionState |= MULTI_TRANSACTION_STATE_FAILED;
+			connection->transactionState &= ~MULTI_TRANSACTION_STATE_STARTING;
+			ReportResultError(connection, result, WARNING);
+		}
+		else
+		{
+			connection->transactionState &= ~MULTI_TRANSACTION_STATE_STARTING;
+			connection->transactionState |= MULTI_TRANSACTION_STATE_STARTED;
+		}
+
+		PQclear(result);
+
+		ForgetResults(connection);
+
+		Assert(PQtransactionStatus(connection->conn) == PQTRANS_INTRANS);
+	}
+}
+
+
+void
+AdjustRemoteTransactionState(MultiConnection *connection)
+{
+	if (InCoordinatedTransaction())
+	{
+		List *connectionList = list_make1(connection);
+
+		AdjustRemoteTransactionStates(connectionList);
+		list_free(connectionList);
+	}
+}
+
+/* -------------------------------------------------------------------------
+ * Execution related functions.
+ * -------------------------------------------------------------------------
+ */
+
+/*
+ * Execute commands on placements in parallel.
+ *
+ * TODO: Use less than one one connection per placement.
```

Comment:
On 2016-09-08 06:32:25 -0700, Marco Slot wrote:
> > +		List *connectionList = list_make1(connection);
> > +
> > +		AdjustRemoteTransactionStates(connectionList);
> > +		list_free(connectionList);
> > +	}
> > +}
> > +
> > +/* -------------------------------------------------------------------------
> > + * Execution related functions.
> > + * -------------------------------------------------------------------------
> > + */
> > +
> > +/*
> > + * Execute commands on placements in parallel.
> > + *
> > + * TODO: Use less than one one connection per placement.
> 
> Any thoughts on how the connection pool is going to be represented in the API, and how we will assign a placement to a connection?

First off: I'm not sure at all that the API is correct here. I
especially dislike that we rely on all placements following each other
on some callsites, but don't enforce that anywhere.  I've not really
found anything I like that's still simple though.

What I'm thinking of doing inside the function, to limit the number of
connections, is to sort the placements so they're grouped by the host,
and then have multiple connections work on the seperate lists. Basically
just process each of the groups, assigning commands to free connections
for that remote node.  I was actually thinking that that shouldn't be
visible to the outside at all.